opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -426,309 +381,209 @@ namespace Opm {
             typedef Dune::BCRSMatrix <MatrixBlockType>      Mat;
             typedef Dune::BlockVector<VectorBlockType>      BVector;
 
-            typedef Dune::MatrixAdapter<Mat,BVector,BVector> Operator;
             auto& ebosJac = ebosSimulator_.model().linearizer().matrix();
             auto& ebosResid = ebosSimulator_.model().linearizer().residual();
 
-            Operator opA(ebosJac);
+            typedef WellModelMatrixAdapter<Mat,BVector,BVector, ThisType> Operator;
+            Operator opA(ebosJac, const_cast< ThisType&  > (*this));
             const double relax = 0.9;
             typedef Dune::SeqILU0<Mat, BVector, BVector> SeqPreconditioner;
             SeqPreconditioner precond(opA.getmat(), relax);
             Dune::SeqScalarProduct<BVector> sp;
 
+            // apply well residual to the residual.
+            wellModel().apply(ebosResid);
```

Comment:
Yes it is. It happens in the WellModelMatrixAdapter. By putting it in the MatrixAdapter you avoid matrix-matrix multiplications and copying of matrices. Again the speed up compared with the earlier version Frankenstein is significant.